include/range/v3/span.hpp

```diff
@@ -0,0 +1,256 @@
+            }
+        }
+        /// \endcond
+
+        template<typename T, std::ptrdiff_t N = dynamic_extent>
+        class span
+          : private box<meta::if_c<N == dynamic_extent,
+                mutable_<std::ptrdiff_t>, constant<std::ptrdiff_t, N>>>
+          , public view_interface<span<T, N>,
+                (N == dynamic_extent ? finite : static_cast<cardinality>(N))>
+        {
+            CONCEPT_ASSERT(N == dynamic_extent || N >= 0);
+            using size_base_t = box<meta::if_c<N == dynamic_extent,
+                mutable_<std::ptrdiff_t>, constant<std::ptrdiff_t, N>>>;
+            T* ptr_ = nullptr;
+
+        public:
+            using element_type = T;
+            using difference_type = std::ptrdiff_t;
+            using index_type = std::ptrdiff_t;
+            using pointer = T*;
+            using reference = T&;
+            using iterator = T*;
+            using reverse_iterator = ranges::reverse_iterator<iterator>;
+
+            static constexpr index_type extent = N;
+
+            // [span.cons], span constructors, copy, assignment, and destructor
+            constexpr span() = default;
+            constexpr span(std::nullptr_t) noexcept
+              : span{}
+            {}
+            constexpr span(pointer ptr, index_type count) noexcept
+              : size_base_t{(RANGES_ENSURE(count >= 0),
+                    RANGES_ENSURE(N == dynamic_extent || count == N),
+                    count)}
+              , ptr_{(RANGES_ENSURE(count == 0 || ptr != nullptr), ptr)}
+            {}
+            constexpr span(pointer first, pointer last) noexcept
+              : span{first, last - first}
+            {}
+            // EXTENSION: disambiguate initialization from "pointer, small-literal"
+            template<typename Int,
+                CONCEPT_REQUIRES_(SignedIntegral<Int>())>
+            constexpr span(pointer first, Int count) noexcept
+              : span{first, index_type{count}}
+            {}
+
+
+            template<typename Rng>
+            using ConversionConstraint =
+                meta::bool_<SizedRange<Rng>::value && ContiguousRange<Rng>::value &&
+                    detail::is_more_cv_than<T, concepts::ContiguousRange::datum_t<Rng>>::value &&
+                    (N == dynamic_extent ||
+                        range_cardinality<Rng>::value == finite ||
+                        range_cardinality<Rng>::value == N)>;
+
+            template<typename Rng,
+                CONCEPT_REQUIRES_(ConversionConstraint<Rng>())>
+            constexpr span(Rng && rng)
+              : span{ranges::data(rng), static_cast<index_type>(ranges::size(rng))}
+            {}
+
+            // [span.sub], span subviews
+            template<std::ptrdiff_t Count,
+                CONCEPT_REQUIRES_(Count >= 0 && (N == dynamic_extent || Count <= N))>
+            constexpr span<T, Count> first() const noexcept
```

Comment:
I hate the naming here. This should be `take`, and `last` should be `drop` (and I know that changes the meaning of the parameter).